internal/worker/worker.go

```diff
@@ -0,0 +1,300 @@
+package worker
+
+import (
+	"archive/tar"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/google/go-containerregistry/pkg/crane"
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/utils"
+)
+
+type Worker interface {
+	LoadKmod(ctx context.Context, cfg *kmmv1beta1.ModuleConfig) error
+	ReloadKmod(ctx context.Context, oldCfg, newCfg *kmmv1beta1.ModuleConfig) error
+	UnloadKmod(ctx context.Context, cfg *kmmv1beta1.ModuleConfig) error
+}
+
+type worker struct {
+	ip     ImagePuller
+	logger logr.Logger
+	mr     ModprobeRunner
+}
+
+func NewWorker(ip ImagePuller, mr ModprobeRunner, logger logr.Logger) Worker {
+	return &worker{
+		ip:     ip,
+		logger: logger,
+		mr:     mr,
+	}
+}
+
+func (w *worker) LoadKmod(ctx context.Context, cfg *kmmv1beta1.ModuleConfig) error {
+	imageName := cfg.ContainerImage
+
+	w.logger.Info("Pulling image", "name", imageName)
+
+	pr, err := w.ip.PullToDisk(ctx, imageName, cfg.InsecurePull)
+	if err != nil {
+		return fmt.Errorf("could not pull %q: %v", imageName, err)
+	}
+
+	if inTree := cfg.InTreeModuleToRemove; inTree != "" {
+		w.logger.Info("Unloading in-tree module", "name", inTree)
+
+		if err = w.mr.Run(ctx, "-rv", inTree); err != nil {
+			return fmt.Errorf("could not remove in-tree module %s: %v", inTree, err)
+		}
+	}
+
+	// TODO copy firmware
```

Comment:
This is something that we need to discuss a little bit more, especially with the customers/partners. I think that at least some of them would like to separate the firmware image from the kmod image, so that they can manage the upgrades better

Reply:
I think we can implement that in a subsequent version. Right now I would focus on feature parity with v1.